RFC: Support for pack.toml (#189)
buildpacks/rfcs: text/0000-pack-toml.md

# Meta
[meta]: #meta
- Name: Support for `pack.toml`
- Start Date: 2021-11-16
- Author(s): [jkutner](http://github.com/jkutner)
- Status: Draft
- RFC Pull Request: (leave blank)
- CNB Pull Request: (leave blank)
- CNB Issue: (leave blank)
- Supersedes: N/A

# Summary
[summary]: #summary

This is a proposal for a `pack.toml` file that is a special case of the `project.toml` descriptor file. The `pack.toml` will include configuration options specific to the Pack CLI, which may not be honored by other platforms.

# Definitions
[definitions]: #definitions

- *project descriptor* - a extension spec defining the [`project.toml`](https://github.com/buildpacks/spec/blob/main/extensions/project-descriptor.md) file.

# Motivation
[motivation]: #motivation

The `project.toml` file has been a useful and essential tool for buildpack users who need to codify certain arguments of their builds. However, there has been much debate around how to handle configuration options that may not be supported by all platforms.

We desire a descriptor file that is:

> * specific to pack (i.e. not platform-neutral)

Author: Two things:
* ability to support pack specific configuration w/o putting it in the project.toml spec
* a file that advertises that a repo can be used with `pack` (like a `Dockerfile` does for docker)

* advertises that the repo containing the file can be built with pack

> * adheres to the best practices and learnings from [`project.toml`](https://github.com/buildpacks/spec/blob/main/extensions/project-descriptor.md)

Reviewer: What do you consider to be the best practices or learnings from `project.toml`?

# What it is
[what-it-is]: #what-it-is

The Pack CLI will support a new descriptor file, `pack.toml`, that is a superset of the project descriptor.
* An app MAY have a `pack.toml`
* An app MAY have a `project.toml`

> * An app MAY have both a `pack.toml` and a `projec.toml`, in which case they will be merged.

Reviewer: ```suggestion
* An app MAY have both a `pack.toml` and a `project.toml`, in which case they will be merged.
```

> * An app MAY have both a `pack.toml` and a `projec.toml`, in which case they will be merged.

Reviewer: What is the use case for having both? Wasn't the purpose of https://github.com/buildpacks/rfcs/blob/main/text/0084-project-descriptor-domains.md to allow for different platform-specific fields to exist in a single master file?

I can't help but feeling like the existence of this proposal is a sign that we need to stop and reevaluate `project.toml`. Clearly being to commit config for `pack` is valuable. But requiring all platforms to conform to the same interface isn't working, even for the simplest platforms like pack.

Reviewer 2: I'd be interested in far it seems impossible to get to a common descriptor that all platforms can adhere to? Actually, I'd expect that at least a lot of those could even be provided by a shared implementation. Isn't that what the RFC discussion about a `prepare` lifecycle is about?

If we need platform specific configuration, and `builder` might be such a case as platforms like kpack or project Piper's `cnbBuild` won't read the `project.toml` until well after running in a process on a builder image, platform specific or optionally interpreted keys in a common file sound far superior to separate files that need merging.

> * An app MAY have both a `pack.toml` and a `projec.toml`, in which case they will be merged.

Reviewer 3: The more I think about it, the more I feel like we shouldn't merge the contents. It's hard to implement things like "removals", "resets", and "append" vs "override". Additionally, by not doing a merging them I feel like it would also be easier for users to comprehend how they are applied. For example, "If `pack.toml` is present it uses that; otherwise it looks for a `project.toml`."

> * An app MAY have both a `pack.toml` and a `project.toml`, in which case the `project.toml` will be ignored.

Reviewer 2: This contradicts https://github.com/buildpacks/rfcs/pull/189/files#diff-9110fb10fa3ce5f9676262392e4b1cb227092e9df99be33edab24508dd1a81a7R73

Author: fixed

> * An app MAY have both a `pack.toml` and a `project.toml`, in which case the `project.toml` will be ignored.

Reviewer 4: ```suggestion
* An app MAY have both a `pack.toml` and a `project.toml` where they will be merged and used.
```

Author: @hone we've gone back and forth on this. Merging could actually be very complicated, and if we support merging we should only do it at the top level shared key (i.e. we're not going to try and merge lists of things for example)

Reviewer 2: @jkutner this is precisely why I have a problem with a platform specific descriptor covering platform agnostic configuration.

I would want to have as much as possible in an agnostic way, at least from the perspective of all platforms other than `pack`. Why? Because for 99% of the non `pack` platforms I can imagine, `pack` will be the platform for local execution. See kpack and our integration into project Piper as two examples and Tekton most likely as a third.

Also, as I pointed out in https://github.com/buildpacks/rfcs/pull/189#discussion_r1038109057 the “`project.toml` will be ignored” statement contradicts other parts of this proposal.

> The `pack.toml` will support an additional `[io.buildpacks.pack]` table that is NOT defined in the [project descriptor spec](https://github.com/buildpacks/spec/blob/main/extensions/project-descriptor.md). The `[io.buildpacks.pack]` table will have the following schema:

Reviewer 3: @jkutner IIRC, we discussed minimizing this by now using reverse name structure but instead having everything in a top-level `[pack]` dictionary. Is that your understanding as well?

```toml
[io.buildpacks.pack]

> builder = "<string (optional)>"

Reviewer 5: `io.buildpacks.builder` already seems to be supported in `project.toml` here: https://github.com/buildpacks/spec/blob/main/extensions/project-descriptor.md#iobuildpacksbuilder-optional. Is this different somehow?

Author: I think that may have been added since I first wrote this. I'll update with a different example that shows something that's specific to pack (like `tag` or maybe `env-file`)

```

Where:

* `builder` is a uri to a CNB Builder docker image

## Example

```
[io.buildpacks.build]
exclude = ["/README.md"]

[[io.buildpacks.group]]
id = "example/java"
version = "1.0.0"

[io.buildpacks.pack]
builder = "example/my-builder-image"
```

# How it Works
[how-it-works]: #how-it-works

The Pack CLI will honor both the `pack.toml` and `project.toml` file descriptors with the following rules:

> * If one or more descriptors are provided on the command line they will all be merged and used.

Reviewer 3: I wonder if this would be simpler if we only had one flag, say something like `--config` that could accept either `pack.toml` OR `project.toml`.

The same could then apply to an inferred discovery mechanism where if the `--config` flag is not presented then it's looked up by precedence and NOT merged.

1. Look up `pack.toml`
2. Look up `project.toml`

* If a `pack.toml` file exists in the workspace and no descriptors are provided on the command line, it will be used.

> * If a `project.toml` file exists in the workspace and no descriptors are provided on the command line, it will be merged with any `pack.toml` that is present in the workspace and the result will be used.

Reviewer 2: I guess I have expressed multiple times, that I don’t like this resolution of the inconsistency in the RFC ;)

* Else no project descriptor will be used

When multiple descriptors are _merged_ the following precedence is given:

* Configuration in a file provided on the command line will take precedence over files provided on the command line after it.

> * Configuration in a `pack.toml` will take precedence over configuration in a `project.toml`

Reviewer 6: I would like some outputs from `pack` to indicate the config being used (even better if it could trace it back to the origin and say why). That way users can know at a glance if their changes have been picked up or if they're not being picked up, get a hint that maybe another file is taking precedence. 

I'm unsure whether that desire can be encoded in the RFC in a sensible way.
